File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/bitonic_array_maximum.cpp
#include <algorithm>
#include <iostream>
#include <list>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

// O(logn) / O(1)
int bitonic_array_maximum(const vector<int> &a) {
  int low = 0, high = a.size() - 1;

  while (low < high) {
    int mid = low + (high - low) / 2;

    if (a[mid] < a[mid + 1]) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }

  return a[low];
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    cout << bitonic_array_maximum(a) << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/BitonicArrayMaximum.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class BitonicArrayMaximum {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        System.out.println(bitonicArrayMaximum(a, 0, n - 1));
      }
    }
  }

  // O(logn) / O(1)
  public static int bitonicArrayMaximum(int[] a, int low, int high) {
    while (low < high) {
      int mid = low + (high - low) / 2;

      if (a[mid] > a[mid + 1]) {
        high = mid;
      } else {
        low = mid + 1;
      }
    }

    return a[low];
  }
}